src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+  """
+
+  potential_value_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                               PotentialValueFn_t] = struct.field(
+                                   pytree_node=False
+                               )
+
+  potential_gradient_fn: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                                  PotentialGradientFn_t] = struct.field(
+                                      pytree_node=False
+                                  )
+
+
+class PotentialModelWrapper(nn.Module):
+  r"""Wrapper class for the neural models.
+
+  Implements a potential value or a vector field.
+  """
+
+  is_potential: bool  # Implements a potential value or a vector field.
+  add_l2_norm: bool  # If true, l2 norm is added to the potential.
+  model: nn.Module  # The potential model.
+
+  def apply(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray], x: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Apply the potential model, optionally add squared Euclidean norm."""
+    z: jnp.ndarray = self.model.apply({"params": params}, x)
+
+    if self.is_potential:
+      z = z.squeeze()
+
+    if self.is_potential and self.add_l2_norm:
+      z = z + 0.5 * jnp.dot(x, x)
+    if not self.is_potential and self.add_l2_norm:
+      z = z + x
+
+    return z
+
+  def potential_value_fn(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray]
+  ) -> PotentialValueFn_t:
+    """Return a function giving the value of the potential."""
+    assert (
+        self.is_potential is True
+    ), "A model should be potential to get potential_value_fn."
+    return lambda x: self.apply(params, x)
+
+  def potential_gradient_fn(
+      self, params: frozen_dict.FrozenDict[str, jnp.ndarray]
+  ) -> PotentialGradientFn_t:
+    """A vector function or gradient of the potential."""
+    if self.is_potential:
+      return jax.grad(self.potential_value_fn(params))
+    return lambda x: self.apply(params, x)
+
+  def create_train_state(
+      self,
+      rng: jax.Array,
+      optimizer: optax.GradientTransformation,
+      input: Union[int, Tuple[int, ...]],
+      **kwargs: Any,
+  ) -> PotentialTrainState:
+    """Create initial training state."""
+    params = self.model.init(rng, jnp.ones(input))["params"]
+
+    return PotentialTrainState.create(
+        apply_fn=self.apply,
+        params=params,
+        tx=optimizer,
+        potential_value_fn=self.potential_value_fn,
+        potential_gradient_fn=self.potential_gradient_fn,
+        **kwargs
+    )
+
+
+class MLP(nn.Module):
+  """A simple MLP model of a potential used in default initialization."""
+
+  dim_hidden: Sequence[int]
+  act_fn: Callable[[jnp.ndarray], jnp.ndarray] = jax.nn.elu
+
+  @nn.compact
+  def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
+    """Apply MLP transform."""
+    for feat in self.dim_hidden[:-1]:
+      x = self.act_fn(nn.Dense(feat)(x))
+    return nn.Dense(self.dim_hidden[-1])(x)
+
+
+class ExpectileNeuralDual(Generic[CostFn_t]):
```

Comment:
Why the use of generics? Imho not necessary, would remove.

Reply:
Generic removed.